### PR TITLE
fix(coverage): add  new option "--ir-minimum" to resolve the "stack too deep" error

### PR DESCRIPTION
**Pull request by spockP**
## Motivation

Partially fix https://github.com/foundry-rs/foundry/issues/3357.  

More and more people (including myself) are using the magical `via-ir` to solve the "stack too deep" issue. However, this would cause problems in test coverage as all the optimizations are disabled in test coverage.  

The solidity team suggests working around this issue by enabling "via-ir" with a ["minimum amount of optimization."](https://github.com/ethereum/solidity/issues/12533#issuecomment-1013073350)  

And this workaround is also recommended in the [FAQ](https://github.com/sc-forks/solidity-coverage/blob/89882a211e9f69ed6b1afc99b2bb9d56b857e686/docs/faq.md?plain=1#L85) of solcoverage.  

## Solution

In this pr, a new "--ir-minimum" option is added to the coverage subcommand, which enables "via-ir" with a minimum amount of optimization. 

## test case 

The contract `Deep` in  `deep.sol` will not compile without any optimizations due to the "stack too deep" error: 

```solidity 

// SPDX-License-Identifier: UNLICENSED
pragma solidity ^0.8.13;

contract Deep {
    mapping(uint256 => uint256) private numbers;

    function sum() public view returns (uint256) {
        uint256 x0 = numbers[0];
        uint256 x1 = numbers[1];
        uint256 x2 = numbers[2];
        uint256 x3 = numbers[3];
        uint256 x4 = numbers[4];
        uint256 x5 = numbers[5];
        uint256 x6 = numbers[6];
        uint256 x7 = numbers[7];
        uint256 x8 = numbers[8];
        uint256 x9 = numbers[9];
        uint256 x10 = numbers[10];
        uint256 x11 = numbers[11];
        uint256 x12 = numbers[12];
        uint256 x13 = numbers[13];
        uint256 x14 = numbers[14];
        uint256 x15 = numbers[15];
        uint256 x16 = numbers[16];
        uint256 x17 = numbers[17];

        // return sum of all numbers
        return x0 + x1 + x2 + x3 + x4 + x5 + x6 + x7 + x8 + x9 + x10 + x11 + x12 + x13 + x14 + x15 + x16 + x17;
    }
}
```

And the test script is as follows: 

```solidity
// SPDX-License-Identifier: UNLICENSED
pragma solidity ^0.8.13;

import "forge-std/Test.sol";
import "../src/deep.sol";

contract TestA is Test {
    Deep sut;

    function setUp() external {
        sut = new Deep();
    }

    function test_GetA() external {
        uint256 result = sut.sum();
        assertEq(result, 0);
    }
}
```

When you run test coverage without "--ir-minimum", you will see the "stack too deep" error:  

```bash 
spock@local:~/case$ forge coverage 
[⠆] Compiling...
[⠊] Compiling 19 files with 0.8.20
[⠢] Solc 0.8.20 finished in 3.67s
Error: 
Compiler run failed:
Error: Compiler error (/solidity/libsolidity/codegen/LValue.cpp:56):Stack too deep. Try compiling with `--via-ir` (cli) or the equivalent `viaIR: true` (standard JSON) while enabling the optimizer. Otherwise, try removing local variables.
  --> src/deep.sol:28:61:
   |
28 |         return x0 + x1 + x2 + x3 + x4 + x5 + x6 + x7 + x8 + x9 + x10 + x11 + x12 + x13 + x14 + x15 + x16 + x17;
   |        
```

When you run test coverage with "--ir-minimum", the "stack too deep" error is gone: 

```bash 
spock@local:~/case$ forge coverage --ir-minimum  
[⠑] Compiling...
[⠢] Compiling 19 files with 0.8.20
[⠰] Solc 0.8.20 finished in 3.53s
Compiler run successful!
Analysing contracts...
Running tests...
| File         | % Lines         | % Statements    | % Branches    | % Funcs       |
|--------------|-----------------|-----------------|---------------|---------------|
| src/deep.sol | 100.00% (19/19) | 100.00% (20/20) | 100.00% (0/0) | 100.00% (1/1) |
| Total        | 100.00% (19/19) | 100.00% (20/20) | 100.00% (0/0) | 100.00% (1/1) |
```

### CODE REVIEW

**Review by Evalir — CHANGES_REQUESTED** [review]
I like this—but we might want to add a CLI warning that adding this flag overrides several configuration options and should probably only be used as a workaround for stack too deep, and link to the relevant issue.

wdyt @mattsse ?

File: cli/src/cmd/forge/coverage.rs [hunk 1]
```diff
@@ -96,11 +106,35 @@ impl CoverageArgs {
...
+                        optimizer_steps: Some("u".to_string()),
+                    }),
+                });
+                project.solc_config.settings.via_ir = Some(true);
+            } else {
+                project.solc_config.settings.optimizer.disable();
+                project.solc_config.settings.optimizer.runs = None;
+                project.solc_config.settings.optimizer.details = None;
+                project.solc_config.settings.via_ir = None;
+            }
```
Comment (mattsse) [inline cli/src/cmd/forge/coverage.rs:137]: so solc --ir-minimum is not an extra setting but instead a profile for the existing settings?

if so then I'd like to move this to `ethers-solc`

**Review by mattsse — CHANGES_REQUESTED** [review]

File: cli/src/cmd/forge/coverage.rs (hunk 1 above)
Comment (spockP) [reply cli/src/cmd/forge/coverage.rs:137]: Gonna add a function to `Settings` in `ethers-solc`.

Comment (Evalir) [reply cli/src/cmd/forge/coverage.rs:137]: You should now be able to do `cargo update -p ethers` and get the latest master, to use the new function!

**Comment by mattsse** [comment]
should be unblocked now

File: Cargo.toml [hunk 2]
```diff
@@ -72,7 +72,7 @@ ethers-middleware = { git = "https://github.com/gakonst/ethers-rs", default-feat
 ethers-etherscan = { git = "https://github.com/gakonst/ethers-rs", default-features = false }
 ethers-solc = { git = "https://github.com/gakonst/ethers-rs", default-features = false }
 
-solang-parser = "=0.3.0"
+solang-parser = "=0.3.1"
```
Comment (spockP) [inline Cargo.toml:75]: I bump solang-parser due to the change of it in [ethers-solc](https://github.com/gakonst/ethers-rs/blob/master/ethers-solc/Cargo.toml#L31). 

However, this introduces some compilation errors resulting in some minor modifications in chisel and fmt.

File: chisel/src/solidity_helper.rs [hunk 3]
```diff
@@ -60,7 +60,6 @@ impl SolidityHelper {
         let mut comments = Vec::with_capacity(DEFAULT_COMMENTS);
         let mut errors = Vec::with_capacity(5);
         let mut out = Lexer::new(input, 0, &mut comments, &mut errors)
-            .flatten()
             .map(|(start, token, end)| (start, token.style(), end))
```
Comment (spockP) [inline chisel/src/solidity_helper.rs:63]: We don't need `flatten`. 

In `solang-parser`, the implementation for `Iterator` trait changes from ([0.3.0](https://github.com/hyperledger/solang/blob/53172f3ad3627b5281086c1ee6f13a57a512f346/solang-parser/src/lexer.rs#L744)): 

```rust
fn next(&mut self) -> Option<Result<'input>>
```
to ([0.3.1](https://github.com/hyperledger/solang/blob/v0.3.1/solang-parser/src/lexer.rs#L745)): 

```rust
fn next(&mut self) -> Option<Spanned<'input>> {
```

File: cli/src/cmd/forge/coverage.rs [hunk 4]
```diff
@@ -96,11 +104,41 @@ impl CoverageArgs {
...
+                // print warning message
+                p_println!(!self.opts.silent => "{}",
+                Paint::yellow(
+                concat!(
+                "Warning! \"--ir-minimum\" flag enables viaIR with minimum optimization, which can result in inaccurate source mappings.\n",
+                "Only use this flag as a workaround if you are experiencing \"stack too deep\" errors.\n",
+                "Note that \"viaIR\" is only available in Solidity 0.8.13 and above.\n",
+                "See more:\n",
+                "https://github.com/foundry-rs/foundry/issues/3357\n"
+                )));
```
Comment (spockP) [inline cli/src/cmd/forge/coverage.rs:128]: @Evalir  CLI warning added

File: cli/src/cmd/forge/coverage.rs [hunk 5]
```diff
@@ -96,11 +104,41 @@ impl CoverageArgs {
...
 
-            // Disable the optimizer for more accurate source maps
-            project.solc_config.settings.optimizer.disable();
-            project.solc_config.settings.optimizer.runs = None;
-            project.solc_config.settings.optimizer.details = None;
-            project.solc_config.settings.via_ir = None;
+            if self.ir_minimum {
+                // TODO: How to detect solc version if the user does not specify a solc version in
+                // config  case1: specify local installed solc ?
+                //  case2: mutliple solc versions used and  auto_detect_solc == true
```
Comment (spockP) [inline cli/src/cmd/forge/coverage.rs:110]: `vir-ir` and `ir-minimum` will only work in solidity 0.8.13 and above. 
I don't find a better way to reject the unsatisfied compilers.

File: fmt/src/solang_ext/ast_eq.rs [hunk 6]
```diff
@@ -395,7 +395,7 @@ derive_ast_eq! { U256 }
 derive_ast_eq! { struct Identifier { loc, name } }
 derive_ast_eq! { struct HexLiteral { loc, hex } }
 derive_ast_eq! { struct StringLiteral { loc, unicode, string } }
-derive_ast_eq! { struct Parameter { loc, ty, storage, name } }
+derive_ast_eq! { struct Parameter { loc, annotation, ty, storage, name } }
```
Comment (spockP) [inline fmt/src/solang_ext/ast_eq.rs:398]: `solang-paser` added a new field to `Parameter`

File: Cargo.toml (hunk 2 above)
Comment (mattsse) [reply Cargo.toml:75]: is updated on master:

https://github.com/foundry-rs/foundry/blob/044f680a5c2541b0fdbd69eb8f33abd57e161569/Cargo.toml#L75-L75

Comment (spockP) [reply Cargo.toml]: @mattsse  rebased.

File: cli/src/cmd/forge/coverage.rs [hunk 7]
```diff
@@ -96,11 +104,41 @@ impl CoverageArgs {
...
+                if let Some(SolcReq::Version(version)) = &config.solc {
+                    if *version < Version::new(0, 8, 13) {
+                        return Err(eyre::eyre!(
+                            "viaIR with minimum optimization is only available in Solidity 0.8.13 and above."
```
Comment (Evalir) [inline cli/src/cmd/forge/coverage.rs:114]: great!

**Review by Evalir — APPROVED** [review]
looks good now! thanks!

File: cli/src/cmd/forge/coverage.rs (hunk 5 above)
Comment (mattsse) [inline cli/src/cmd/forge/coverage.rs:110]: I think we can ignore this this,

all options are sanitized in ethers-solc depending on version, meaning unsupported options are removed before calling solc

**Review by mattsse — APPROVED** [review]
tysm for this

**Comment by davidlucid** [comment]
Hey guys! I have I seem to have found a bug with `forge coverage --ir-minimum`. When running `forge coverage` alone, I get a "stack too deep" error; when running `forge coverage --ir-minimum`, it works, but it tells me I don't have coverage on an area of my tests that I clearly have coverage for (specifically, 3 lines of assembly code for loading data into memory). I know I have coverage for these 3 lines because if you remove some of the contract files and one of the test files, the normal `forge coverage` command works and tells me I have 100% coverage (but again tells me I don't have 100% coverage if I run `forge coverage --ir-minimum`). Any ideas on what to do here to get coverage to show 100%? Thanks! @mattsse @Evalir @spockP

**Pull request merged**